### PR TITLE
Build binaries for both Linux and MacOS, add tests to the CI and do not rely on your system libyara version

### PR DESCRIPTION
**Pull request by macbre**
* https://github.com/springmeyer/node-addon-example
* https://github.com/mapbox/node-pre-gyp
* use [v3.9 of `libyara`](https://github.com/VirusTotal/yara/archive/refs/tags/v3.9.0.tar.gz) - revert 61782927f0f3c6f6d5ea9e2c7bf529fe95f538c5 so that we can always fetch the specific libyara version (for instance Mac's `brew` only has 3.10 which does not work with this package)
* have all pre-build binaries uploaded to the GH release ([an example](https://github.com/vercel/pkg-fetch/releases/tag/v3.4))

Have something similar to what `sqlite3` package does:

```
> sqlite3@5.0.2 install /home/runner/work/jetpack-backups/jetpack-backups/node_modules/sqlite3
> node-pre-gyp install --fallback-to-build
node-pre-gyp WARN Using request for node-pre-gyp https download 
[sqlite3] Success: "/home/runner/work/jetpack-backups/jetpack-backups/node_modules/sqlite3/lib/binding/napi-v3-linux-x64/node_sqlite3.node" is installed via remote
```

Completed so far:

* [x] build binaries are upload as GH Action run artifacts
* [x] added Mocha - tests are now executed via `npm t` 

---

See my comment in https://github.com/Automattic/jetpack-backups/pull/1959#issuecomment-889137819

### CODE REVIEW

**Comment by miguelxpn** [comment]
This is great, I apologize for missing it all this time. Not only this will speed up our builds, but it will also make it easier to migrate to newer versions potentially. I'll give it a spin this weekend and see how we can incorporate this to the migration plan which hopefully will be happening pretty soon! Thanks again for working on this!

**Comment by macbre** [comment]
We can also use `gh` CLI tool to both create releases and also upload binaries there (for where we can fetch them on `npm i`).

```
  Use automatically generated release notes
  $ gh release create v1.2.3 --generate-notes
  
  Use release notes from a file
  $ gh release create v1.2.3 -F changelog.md
  
  Upload all tarballs in a directory as release assets
  $ gh release create v1.2.3 ./dist/*.tgz
```

**Comment by macbre** [comment]
https://github.com/Automattic/node-yara/actions/runs/2549470791:

Screenshot 2022-06-23 at 14 37 44

**Comment by miguelxpn** [comment]
> https://github.com/Automattic/node-yara/actions/runs/2549470791:
> 
> Screenshot 2022-06-23 at 14 37 44
> 
> These files should basically land here as `yara.node`:
> 
> ```
> vaultpress@macbre.dca.vaultpress.com:/home/jetpack-backups$ ls node_modules/yara/build/Release/                           
> obj.target  yara.node
> ```
> 
> @miguelxpn - how can I test yara-powered features on my sandbox? Which `bbin` script should I run?

You can just use the scanner.js script to scan any local file.

**Comment by miguelxpn** [comment]
> @miguelxpn: tests from this repo segfault on MacOS (they pass on Ubuntu):
> 
> ```
> npm test || true
> > @automattic/yara@2.3.0 test /Users/runner/work/node-yara/node-yara
> > mocha test/*
>   index.js
>     Scanner.configure()
>       ✔ call scan() before configure()
>       ✔ rules.string - missing is ignored
>       ✔ rules.string - empty is ignored
> Segmentation fault: 11
> ```

Found out the issue, the compiler callback can return Null for the file_name string as sometimes you don't pass rules as a file but as as in memory string, I added a commit here handling that case, it seems some tests are still failing on Mac though, will keep looking at them to see what other divergences we may have.

**Comment by miguelxpn** [comment]
> > @miguelxpn: tests from this repo segfault on MacOS (they pass on Ubuntu):
> > ```
> > npm test || true
> > > @automattic/yara@2.3.0 test /Users/runner/work/node-yara/node-yara
> > > mocha test/*
> >   index.js
> >     Scanner.configure()
> >       ✔ call scan() before configure()
> >       ✔ rules.string - missing is ignored
> >       ✔ rules.string - empty is ignored
> > Segmentation fault: 11
> > ```
> 
> Found out the issue, the compiler callback can return Null for the file_name string as sometimes you don't pass rules as a file but as as in memory string, I added a commit here handling that case, it seems some tests are still failing on Mac though, will keep looking at them to see what other divergences we may have.

The way we were deserializing the filename it was probably causing a dangling pointer that caused different undefined behavior on Mac and Linux, tests are passing on both platforms now.

**Comment by macbre** [comment]
Thanks @miguelxpn for fixing tests. I've now made passing `npm test` the required step in the CI.

File: src/yara.cc [hunk 1]
```diff
@@ -519,7 +520,11 @@ void compileCallback(int error_level, const char* file_name, int line_number,
 	CompileArgs* args = (CompileArgs*) user_data;
 
 	std::ostringstream oss;
-	oss << args->rule_config->index << ":" << line_number << ":" << message << ":" << file_name;
+	oss << args->rule_config->index << ":" << line_number << ":" << message << ":";
+
+	if (file_name != NULL) {
+		oss << file_name;
+	}
```
Comment (macbre) [inline src/yara.cc:527]: @miguelxpn - should we backport these fixes to the `master` branch?

Comment (miguelxpn) [reply src/yara.cc:527]: I just cherry picked them into master!

Comment (macbre) [reply src/yara.cc:527]: Perfect, thanks!

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -10,26 +10,44 @@ on:
...
+        npm install --build-from-source
+
+        ls -lh ./build/Release
+        md5sum ./build/Release/yara.node || md5 ./build/Release/yara.node || true
+
+    - name: Upload binaries as Action run artifacts
+      uses: actions/upload-artifact@v2
+      with:
+        name: node${{ matrix.node-version }}-${{ matrix.os }}-yara.node
+        path: ./build/Release/yara.node
```
[the remark this answers is not in the document]
Comment (macbre) [reply .github/workflows/tests.yml:50]: https://docs.github.com/en/actions/using-workflows/storing-workflow-data-as-artifacts

**Comment by macbre** [comment]
A friendly reminder about this PR 🙂  Moving towards binaries provided by the `npm` module itself will provide quite a few benefits (quicker deploys, ability to run unit tests for the root directory locally, ...).

**Review by miguelxpn — APPROVED** [review]
LGTM! Sorry for taking so long to review this, had a hectic few weeks :)

**Comment by miguelxpn** [comment]
Once this is merged I'll use it as a base to upgrade our libyara. Should make the upgrade much easier, thanks @macbre

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+endif
+
+YARA?=3.9.0
+
+libyara: yara
+
+yara:
+	-rm -rf $(BASE)/build/yara
+	-rm -rf $(BASE)/deps/yara-$(YARA)
+	test -f $(BASE)/deps/yara-$(YARA).tar.gz || curl -L -k https://github.com/VirusTotal/yara/archive/v$(YARA).tar.gz > $(BASE)/deps/yara-$(YARA).tar.gz
```
Comment (macbre) [inline Makefile:21]: WIth this PR we'll stop depending on system-wide `libyara` version and just fetch the one defined in `YARA` variable above.

**Pull request merged**